### PR TITLE
[Merged by Bors] - feat: Port Computability.Ackermann

### PR DESCRIPTION
**Pull request by casavaca**
---

Adjusted some proofs (probably) due to [4#2136](https://github.com/leanprover/lean4/issues/2136)

- [x] depends on: #2360 
[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2360
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Computability/Ackermann.lean [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+- `exists_lt_ack_of_primrec`: any primitive recursive function is pointwise bounded above by `ack m`
+  for some `m`.
+- `not_primrec₂_ack`: the two-argument Ackermann function is not primitive recursive.
+
+## Proof approach
+
+We very broadly adapt the proof idea from
+https://www.planetmath.org/ackermannfunctionisnotprimitiverecursive. Namely, we prove that for any
+primitive recursive `f : ℕ → ℕ`, there exists `m` such that `f n < ack m n` for all `n`. This then
+implies that `λ n, ack n n` can't be primitive recursive, and so neither can `ack`. We aren't able
```
Comment (vihdzp) [inline Mathlib/Computability/Ackermann.lean:31]:
```suggestion
implies that `fun n ↦ ack n n` can't be primitive recursive, and so neither can `ack`. We aren't able
```
Might need some line wrapping fixes.

File: Mathlib/Computability/Ackermann.lean [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+to use the same bounds as in that proof though, since our approach of using pairing functions
+differs from their approach of using multivariate functions.
+
+The important bounds we show during the main inductive proof (`exists_lt_ack_of_primrec`) are the
+following. Assuming `∀ n, f n < ack a n` and `∀ n, g n < ack b n`, we have:
+
+- `∀ n, Nat.mkpair (f n) (g n) < ack (max a b + 3) n`.
+- `∀ n, g (f n) < ack (max a b + 2) n`.
+- `∀ n, Nat.rec (f n.unpair.1) (λ (y IH : ℕ), g (Nat.mkpair n.unpair.1 (Nat.mkpair y IH)))
```
Comment (vihdzp) [inline Mathlib/Computability/Ackermann.lean:40]:
```suggestion
- `∀ n, Nat.rec (f n.unpair.1) (fun (y IH : ℕ) ↦ g (Nat.mkpair n.unpair.1 (Nat.mkpair y IH)))
```

File: Mathlib/Computability/Ackermann.lean [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+
+-- Porting note: re-written to get rid of ack_three_aux
+@[simp]
+theorem ack_three (n : ℕ) : ack 3 n = 2 ^ (n + 3) - 3 := by
+  induction' n with n IH
+  · rfl
+  · rw [ack_succ_succ, IH, ack_two]
+    rw [Nat.succ_add, Nat.pow_succ 2 (n + 3), mul_comm _ 2]
+    rw [Nat.mul_sub_left_distrib]
+    rw [←Nat.sub_add_comm, two_mul 3, Nat.add_sub_add_right]
```
Comment (vihdzp) [inline Mathlib/Computability/Ackermann.lean:106]: Can't these `rw` be merged?

**Review by vihdzp — COMMENTED** [review]
Looks good to me.

File: Mathlib/Computability/Ackermann.lean (hunk 3 above)
Comment (Parcly-Taxel) [reply Mathlib/Computability/Ackermann.lean]: Done.

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#2360~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Pull request closed without merging**